src/jrd/MetaName.h

```diff
@@ -0,0 +1,365 @@
+		static unsigned getWordCapacity();
+
+	private:
+		static unsigned getWordLength(FB_SIZE_T len);
+
+#if GROW_DEBUG > 1
+		static const unsigned SEG_BUFFER_SIZE = 256;		// size in sizeof(pointer)
+#else
+		static const unsigned SEG_BUFFER_SIZE = 16384;		// size in sizeof(pointer)
+#endif
+		void* buffer[SEG_BUFFER_SIZE];
+		std::atomic<unsigned> position;
+	};
+	Segment* segment;
+	unsigned segCount;
+
+	Firebird::Mutex mutex;	// The single mutex to protect dictionary when needed
+};
+
+class MetaName
+{
+private:
+	Dictionary::Word* word;
+	static const char* EMPTY;
+
+	void test();
+	Dictionary::Word* get(const char* s, FB_SIZE_T l);
+
+	Dictionary::Word* get(const char* s)
+	{
+		return get(s, s ? fb_strlen(s) : 0);
+	}
+
+public:
+	MetaName()
+		: word(nullptr)
+	{ }
+
+	MetaName(const char* s)
+		: word(get(s))
+	{ }
+
+	MetaName(const Firebird::MetaString& s);
+
+	MetaName(const char* s, FB_SIZE_T l)
+		: word(get(s, l))
+	{ }
+
+	MetaName(const MetaName& m)
+		: word(m.word)
+	{
+		test();
+	}
+
+	MetaName(const Firebird::AbstractString& s)
+		: word(get(s.c_str(), s.length()))
+	{ }
+
+
+	explicit MetaName(MemoryPool&)
+		: word(nullptr)
+	{ }
+
+	MetaName(MemoryPool&, const char* s)
+		: word(get(s))
+	{ }
+
+	MetaName(MemoryPool&, const char* s, FB_SIZE_T l)
+		: word(get(s, l))
+	{ }
+
+	MetaName(MemoryPool&, const MetaName& m)
+		: word(m.word)
+	{
+		test();
+	}
+
+	MetaName(MemoryPool&, const Firebird::AbstractString& s)
+		: word(get(s.c_str(), s.length()))
+	{ }
+
+
+	MetaName& assign(const char* s, FB_SIZE_T l)
+	{
+		word = get(s, l);
+		return *this;
+	}
+
+	MetaName& assign(const char* s)
+	{
+		word = get(s);
+		return *this;
+	}
+
+	MetaName& operator=(const char* s)
+	{
+		word = get(s);
+		return *this;
+	}
+
+	MetaName& operator=(const Firebird::AbstractString& s)
+	{
+		word = get(s.c_str(), s.length());
+		return *this;
+	}
+
+	MetaName& operator=(const MetaName& m)
+	{
+		word = m.word;
+		test();
+		return *this;
+	}
+
+	MetaName& operator=(const Firebird::MetaString& s);
+
+	FB_SIZE_T length() const
+	{
+		return word ? word->length() : 0;
+	}
+
+	const char* c_str() const
+	{
+		return word ? word->c_str() : EMPTY;
+	}
+
+	const char* nullStr() const
+	{
+		return word ? word->c_str() : nullptr;
+	}
+
+	bool isEmpty() const
+	{
+		return !word;
+	}
+
+	bool hasData() const
+	{
+		return word;
+	}
+
+	char operator[](unsigned n) const
+	{
+		fb_assert(n < length());
+		return word->c_str()[n];
+	}
+
+	const char* begin() const
+	{
+		return word ? word->c_str() : EMPTY;
+	}
+
+	const char* end() const
+	{
+		return word ? &word->c_str()[length()] : EMPTY;
+	}
+
+	int compare(const char* s, FB_SIZE_T l) const;
+
+	int compare(const char* s) const
+	{
+		return compare(s, s ? fb_strlen(s) : 0);
+	}
+
+	int compare(const Firebird::AbstractString& s) const
+	{
+		return compare(s.c_str(), s.length());
+	}
+
+	int compare(const MetaName& m) const
+	{
+		return compare(m.begin(), m.length());
```

Comment:
When you're comparing two MetaNames, can't you just compare the word pointer? Or at least, compare it first and if they're equal the bytes does not need to be compared.

Reply:
That's done in operator== and operator!=, but you are right, some optimization is possible here.